### PR TITLE
[MLIR][Linalg] Support dynamic sizes in `lower_unpack`

### PR DESCRIPTION
**Pull request by srcarroll**

### CODE REVIEW

**Comment by srcarroll** [comment]
Not sure what the deal is with there buildkite failure. Seems unrelated to my changes

File: mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp [hunk 1]
```diff
@@ -381,11 +381,6 @@ FailureOr<LowerUnPackOpResult> linalg::lowerUnPack(RewriterBase &rewriter,
     return rewriter.notifyMatchFailure(unPackOp, "outer dims perm NYI");
 
   RankedTensorType packedTensorType = unPackOp.getSourceType();
```
Comment (chelini) [inline mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp:383]: nit: I would move this line after `rewriter.setInsertionPoint(unPackOp);` as it is used further down now.

File: mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp [hunk 2]
```diff
@@ -434,8 +429,21 @@ FailureOr<LowerUnPackOpResult> linalg::lowerUnPack(RewriterBase &rewriter,
       RankedTensorType::Builder(packedTensorType).setShape(stripMinedShape);
   RankedTensorType collapsedType = tensor::CollapseShapeOp::inferCollapsedType(
       stripMinedTensorType, packingMetadata.reassociations);
+
+  // Get dynamic dims from input tensor in order of stripMinedTensor
```
Comment (chelini) [inline mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp:433]: Not sure if I understand the comment. I would rephrase as: `Get dynamic dims from input tensor based on lastDimsToInsertPositionsPerm permutation`.

File: mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp [hunk 3]
```diff
@@ -434,8 +429,21 @@ FailureOr<LowerUnPackOpResult> linalg::lowerUnPack(RewriterBase &rewriter,
...
+  // `tensor.empty` op
+  SmallVector<OpFoldResult, 4> dims =
+      tensor::getMixedSizes(rewriter, loc, unPackOp.getSource());
+  applyPermutationToVector(dims, lastDimsToInsertPositionsPerm);
+  SmallVector<Value, 4> dynDims;
```
Comment (chelini) [inline mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp:438]: We can avoid the loop using another builder method for emptyOp:

```
SmallVector<OpFoldResult> dims =
      tensor::getMixedSizes(rewriter, loc, unPackOp.getSource());
applyPermutationToVector(dims, lastDimsToInsertPositionsPerm);

auto emptyOp = rewriter.create<tensor::EmptyOp>(
      loc, dims, stripMinedTensorType.getElementType());
```

**Review by chelini — COMMENTED** [review]
Thanks!

File: mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp [hunk 4]
```diff
@@ -434,8 +428,14 @@ FailureOr<LowerUnPackOpResult> linalg::lowerUnPack(RewriterBase &rewriter,
       RankedTensorType::Builder(packedTensorType).setShape(stripMinedShape);
   RankedTensorType collapsedType = tensor::CollapseShapeOp::inferCollapsedType(
       stripMinedTensorType, packingMetadata.reassociations);
-  auto emptyOp =
-      rewriter.create<tensor::EmptyOp>(loc, stripMinedTensorType, ValueRange{});
+
+  // Get dynamic dims from input tensor based on lastDimsToInsertPositionsPerm
+  // permutation
```
Comment (chelini) [inline mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp:433]: nit. '.' at the end of the sentence.

File: mlir/test/Dialect/Linalg/transform-lower-pack.mlir [hunk 5]
```diff
@@ -464,6 +464,129 @@ module attributes {transform.with_named_sequence} {
...
+          !transform.op<"linalg.transpose">,
+          !transform.op<"tensor.collapse_shape">,
+          !transform.op<"tensor.extract_slice">)
+      transform.yield
+  }
+}
+
+// -----
+
+// Check that we can lower unpack "as unpad" with dynamic dims
```
Comment (chelini) [inline mlir/test/Dialect/Linalg/transform-lower-pack.mlir:550]: nit. '.' at the end of the sentence.

**Review by chelini — APPROVED** [review]
Thanks, it looks good to me.
